sql: nil pointer dereference in Executor

node 9 on indigo died with the following panic:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: INSERT INTO blocks (block_id, writer_id, block_num, raw_bytes) VALUES (7359107162809420170, 'c9377e9c-a3e2-4996-bde1-8664f15d2621', 4347280, $1): runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: INSERT INTO blocks (block_id, writer_id, block_num, raw_bytes) VALUES (7359107162809420170, 'c9377e9c-a3e2-4996-bde1-8664f15d2621', 4347280, $1): runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: INSERT INTO blocks (block_id, writer_id, block_num, raw_bytes) VALUES (7359107162809420170, 'c9377e9c-a3e2-4996-bde1-8664f15d2621', 4347280, $1): runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1274be4]

goroutine 1410 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc420250280)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198 +0x6e
panic(0x1ad3080, 0xc45069bd10)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc421a7c000, 0x7fe71e37e500, 0xc420b26d20)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:348 +0x74
panic(0x1ad3080, 0xc45069bd10)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc42198eabb, 0x90)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:497 +0x131
panic(0x1afdde0, 0x28fbee0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4203c7040, 0xc420d42000, 0xc42198eabb, 0x90, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:717 +0x734
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4203c7040, 0xc420d42000, 0xc42198eabb, 0x90, 0xc4a588b498, 0x0, 0x0, 0x0, 0xc42198eb00)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:503 +0x180
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc421a7c000, 0xc42198eabb, 0x90, 0xc4a588b498, 0x2da5d70, 0x0, 0x0, 0x1b12600, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:831 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleExecute(0xc421a7c000, 0xc421a7c028, 0xc420f8a2a0, 0xc421a7c045)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:818 +0x172
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc421a7c000, 0x7fe71e37e500, 0xc420b26d20, 0xc420846cc0, 0x5400, 0xc4203a8690, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:450 +0xb37
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4203a8540, 0x7fe71e37e4c0, 0xc420ff9840, 0x2940d80, 0xc420b3aa80, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:415 +0x927
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1(0x2942e20, 0xc420d05900)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:587 +0x153
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc420250280, 0xc42026a020, 0x2942e20, 0xc420d05900, 0xc420a59bc0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x239
```

This seems to be caused by [this line](https://github.com/cockroachdb/cockroach/blob/4d2c8be8e93e560a3a4dabda87c69c5363056311/pkg/sql/executor.go#L734) in `executor.go`.

--- a/pkg/sql/executor.go
+++ b/pkg/sql/executor.go
@@ -729,18 +729,21 @@ func (e *Executor) execRequest(
 			if aErr, ok := err.(*client.AutoCommitError); ok {
 				// TODO(andrei): Until #7881 fixed.
 				{
-					log.Eventf(session.Ctx(), "executor got AutoCommitError: %s\n"+
-						"txn: %+v\nexecOpt.AutoRetry %t, execOpt.AutoCommit:%t, stmts %+v, remaining %+v",
-						aErr, txnState.txn.Proto(), execOpt.AutoRetry, execOpt.AutoCommit, stmts,
-						remainingStmts)
-					if txnState.txn == nil {
+					if txnState.txn != nil {
+						log.Eventf(session.Ctx(), "executor got AutoCommitError: %s\n"+
+							"txn: %+v\nexecOpt.AutoRetry %t, execOpt.AutoCommit:%t, stmts %+v, remaining %+v",
+							aErr, txnState.txn.Proto(), execOpt.AutoRetry, execOpt.AutoCommit, stmts,
+							remainingStmts)
+					} else {
 						log.Errorf(session.Ctx(), "7881: AutoCommitError on nil txn: %s, "+
-							"txnState %+v, execOpt %+v, stmts %+v, remaining %+v",
-							aErr, txnState, execOpt, stmts, remainingStmts)
+							"txnState %+v, execOpt %+v, stmts %+v, remaining %+v, txn captured before executing batch: %+v",
+							aErr, txnState, execOpt, stmts, remainingStmts, txn)
 						txnState.sp.SetBaggageItem(keyFor7881Sample, "sample me please")
 					}
 				}
 				e.TxnAbortCount.Inc(1)
+				// TODO(andrei): Once 7881 is fixed, this should be
+				// txnState.txn.CleanupOnError().
 				txn.CleanupOnError(session.Ctx(), err)
 			}
 		}

